### PR TITLE
partial attempt at stopping  non-differentiable values from being materialized

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110592
* #102809
* #110687
* #110668
* #110591
* #110504
* #110501

### CODE REVIEW

File: torch/_functorch/aot_autograd.py [hunk 1]
```diff
@@ -422,6 +422,8 @@ class OutputAliasInfo:
     base_idx: Optional[int]
     # If it is a Tensor, what the dynamic dims are (otherwise is None)
     dynamic_dims: Optional[Set[int]]
+    # requires_grad
+    requires_grad: bool
```
Comment (bdhirsh) [inline torch/_functorch/aot_autograd.py:426]: nit: `ViewAndMutationMeta.requires_grad_info` already has this info, since it is of length (# mutated inputs + # user outputs).

File: torch/_functorch/aot_autograd.py [hunk 2]
```diff
@@ -969,9 +970,10 @@ def inner(*flat_args):
             for o, info in zip(flat_f_outs, output_info)
             if info.output_type in [OutputType.non_alias, OutputType.unsafe_view_alias, OutputType.custom_function_view]
             and issubclass(info.raw_type, torch.Tensor)
+            and info.requires_grad
```
Comment (bdhirsh) [inline torch/_functorch/aot_autograd.py:973]: I think you can equivalently check this with:
```
for o, info, requires_grad_info in zip(flat_f_outs, output_info, output_requires_grad_info):
...
and requires_grad_info
```

File: torch/_functorch/aot_autograd.py [hunk 3]
```diff
@@ -969,9 +970,10 @@ def inner(*flat_args):
...
         ]
         # intermediate bases are also included in the backward graph
-        f_tangents = f_input_tangents + f_output_tangents + intermediate_bases
```
Comment (bdhirsh) [inline torch/_functorch/aot_autograd.py:974]: hmm, I don't think we can't blindly remove `f_input_tangents` here, because if we have an input that was mutated **and** required gradients, then it needs to participate in the backward.

You should be able to filter `f_input_tangents` the same way you filtered `f_output_tangents` above:
```
        f_input_tangents = [
            inp
            for inp, info, requires_grad_info in zip(flat_f_args, input_info, input_requires_grad_info)
            if info.mutates_data and requires_grad_info
        ]
```

File: torch/_functorch/aot_autograd.py (hunk 1 above)
Comment (Chillee) [reply torch/_functorch/aot_autograd.py:426]: Any reason you prefer to have a separate object rather than have it on here?

Comment (bdhirsh) [reply torch/_functorch/aot_autograd.py:441]: Yeah this is a reasonable q: Let me give context. There are a few other places where we need to distinguish which tensors require gradients:

(1) Affects both mutated inputs and fw outputs ([code](https://github.com/pytorch/pytorch/blob/main/torch/_functorch/aot_autograd.py#L3121)): At runtime, the compiled forward graph returns **both** any updated inputs, and any user fw outs. For both of these groups of tensors, we need to know which ones do not require gradients, so we can mark them as non_differentiable.

(2) Affects just mutated inputs ([code](https://github.com/pytorch/pytorch/blob/main/torch/_functorch/aot_autograd.py#L1177)): During tracing of the joint, if we have a mutated input that requires gradients, we need to clone() it in the forward, so that we can autograd.grad() w.r.t. the input pre-mutation (this is actually kind of sub-optimal since the clone() affects runtime perf and is just to appease the autograd engine, but this case should be rare). 

(3) Affects just fw outputs ([code](https://github.com/pytorch/pytorch/blob/main/torch/_functorch/aot_autograd.py#L2685)): At runtime after our `CompiledFunction.apply()` finishes, we need to regenerate fw outs that alias inputs. We need to know if the output alias no longer requires grad, which implies that a detach() happened during tracing (so we need to reapply that detach at runtime).

If we do what you put above and move the requires_grad info directly on `OutputInfo`, then we'd have to either:

(a) Also put requires_grad info on `InputInfo` objects - **but**, every piece of code in AOTAutograd that care about requires grad information on inputs would have to manually filter down to the input infos that correspond to mutated inputs

(b) Add another piece of metadata that just correponds to "mutated inputs", that we tack requires_grad info on.

Let me know what you think. (a) Actually might make things easier to reason about (even if it's a bit more boilerplate), but it will be a pretty annoying refactor.

**Comment by Chillee** [comment]
Closed in favor of https://github.com/pytorch/pytorch/pull/110721
